service-statistics-pxweb/src/main/java/fi/nls/oskari/control/statistics/plugins/pxweb/parser/PxwebIndicatorsParser.java

```diff
@@ -27,48 +25,206 @@
+                continue;
+            }
+            */
+
+            try {
+                PxTableItem table = getPxTable(path, item.id);
                 StatisticalIndicator ind = new StatisticalIndicator();
                 ind.setId(item.id);
                 ind.addName(PropertyUtil.getDefaultLanguage(), item.text);
-                setupMetadata(ind, path);
-                for(DatasourceLayer layer : layers) {
-                    StatisticalIndicatorLayer l = new StatisticalIndicatorLayer(layer.getMaplayerId(), ind.getId());
-                    l.addParam("baseUrl", url);
-                    ind.addLayer(l);
-                }
+                final StatisticalIndicatorDataModel selectors = getModel(table);
+                ind.setDataModel(selectors);
+                setupLayers(ind, layers, url);
                 indicators.add(ind);
+            } catch (IOException e) {
+                LOG.error(e, "Error getting indicators from Pxweb datasource:", config.getUrl());
             }
+        }
+
+        return indicators;
+    }
 
+    protected List<StatisticalIndicator> parsePxFile(String url, String path, List<DatasourceLayer> layers) {
+        if(config.getIndicatorKey() == null) {
+            LOG.warn("Tried to parse px-file as indicator list but missing indicator key configuration!");
+            return Collections.emptyList();
+        }
+        try {
+            PxTableItem table = getPxTable(path);
+            List<StatisticalIndicator> indicatorList = readPxTableAsIndicators(table);
+            setupLayers(indicatorList, layers, url);
+            return indicatorList;
         } catch (IOException e) {
             LOG.error(e, "Error getting indicators from Pxweb datasource:", config.getUrl());
         }
-        return indicators;
+        return Collections.emptyList();
+    }
+
+    protected List<PxFolderItem> readFolderListing(String url) {
+        try {
+            String jsonResponse = IOHelper.getURL(url);
+            List<PxFolderItem> list =
+                    mapper.readValue(jsonResponse, mapper.getTypeFactory().constructCollectionType(List.class, PxFolderItem.class));
+            return list;
+        } catch (IOException e) {
+            LOG.error(e, "Error getting indicators from Pxweb datasource:", config.getUrl());
+        }
+        return Collections.emptyList();
+    }
+
+    private void setupLayers(List<StatisticalIndicator> list, List<DatasourceLayer> layers, String baseUrl) {
+        for(StatisticalIndicator ind: list) {
+            setupLayers(ind, layers, baseUrl);
+        }
+    }
+
+    private void setupLayers(StatisticalIndicator ind, List<DatasourceLayer> layers, String baseUrl) {
+        for(DatasourceLayer layer : layers) {
+            StatisticalIndicatorLayer l = new StatisticalIndicatorLayer(layer.getMaplayerId(), ind.getId());
+            l.addParam("baseUrl", baseUrl);
+            ind.addLayer(l);
+        }
+    }
+
+    /*
```

Comment:
What is this massive comment block describing? I find it more distracting than describing

Reply:
Yep, this should be a test resource and a test against it would be nice. It was part of the code when I started, just moved it when I rearranged the methods to make the file more readable.